File: script/sceneGraph/Scene.js
define(["sceneGraph/Camera", "sceneGraph/TransformNode", "sceneGraph/visitor/RenderVisitor", "resourceManager/ResourceManager"],
    function (Camera, TransformNode, RenderVisitor, ResourceManager) {
        "use strict";

        /**
         *
         * @param {ResourceManager} ResourceManager
         * @constructor
         * @param gl
         * @param nViewportWidth
         * @param nViewportHeight
         */
        var Scene = function (ResourceManager, gl, nViewportWidth, nViewportHeight) {
            this.gl = gl;
            this.resourceManager = ResourceManager;

            gl.clearColor(0.0, 0.0, 0.0, 1.0);
            gl.enable(gl.DEPTH_TEST);

            this.camera = new Camera(nViewportWidth, nViewportHeight);
            this.root = new TransformNode(function (mMatrix) {
                return mMatrix;
            });

            this.oInstructions = {shaders: []};
        };

        Scene.prototype.refresh = function () {
            var oVisitor = new RenderVisitor();

            oVisitor.process(this.root);

            var aInstructions = oVisitor.instructions;
            var oSorted = {shaders: []};
            for (var i = 0; i < aInstructions.length; i++) {
                var oInstruction = aInstructions[i];

                if (!oSorted[oInstruction.shader.id]) {
                    oSorted[oInstruction.shader.id] = [];
                    oSorted.shaders.push(oInstruction.shader.id);
                }

                oSorted[oInstruction.shader.id].push(oInstruction);
            }

            this.oInstructions = oSorted;
        };

        Scene.prototype.render = function () {
            /* jshint bitwise: false */
            /* jshint loopfunc: true */
            this.gl.viewport(0, 0, this.camera.nWidth, this.camera.nHeight);
            this.gl.clear(this.gl.COLOR_BUFFER_BIT | this.gl.DEPTH_BUFFER_BIT);

            var thi$ = this;
            for (var i = 0; i < this.oInstructions.shaders.length; i++) {
                var sShader = this.oInstructions.shaders[i];
                this.resourceManager.getResource({id: sShader}, ResourceManager.ResourceType.SHADER, function (oShader) {
                    thi$.renderWithShader(oShader, thi$);
                });
            }
        };

        /**
         *
         * @param {Shader} oShader
         */
        Scene.prototype.renderWithShader = function (oShader, thi$) {
            oShader.use(thi$.camera.mProjection);

            var aShaderInstructions = thi$.oInstructions[oShader.id];
            for (var i = 0; i < aShaderInstructions.length; i++) {
                var oInstruction = aShaderInstructions[i];

                oInstruction.object.model.draw(oInstruction.matrix);
            }
        };

        return Scene;
    }
);

File: lib/scene_graph/scene.dart
part of scene_graph;

/**
 * Represents a scene to be rendered.
 */
class Scene {

  gl.RenderingContext context;
  ResourceManager resourceManager;

  /**
   * The root node of the scene graph.
   */
  GraphNode root = new TransformNode((mat4) => mat4);
  Camera camera;

  Map<String, Set<RenderInstruction>> instructions = new Map();

  Scene(ResourceManager resourceManager, gl.RenderingContext context,
        int viewportWidth,
        int viewportHeight) : resourceManager = resourceManager, this.context = context {

    context.clearColor(0, 0, 0, 1);
    context.enable(gl.DEPTH_TEST);

    camera = new Camera(viewportWidth, viewportHeight);
  }

  /**
   * Recalculate the modelview matrix and shader for all objects in the scene.
   */

  void refresh() {
    (this.root as TransformNode).transform = (
        mat4) => this.camera.getViewMatrix();
    this.instructions.clear();

    RenderVisitor visitor = new RenderVisitor(this.context);

    visitor.process(this.root);

    for (RenderInstruction instruction in visitor.instructions) {

      if (!this.instructions.containsKey(instruction.shader.id)) {
        this.instructions[instruction.shader.id] = new Set();
      }

      this.instructions[instruction.shader.id].add(instruction);
    }
  }

  /**
   * Render the scene to screen using instructions calculated by refresh.
   */

  void render() {

    this.context.viewport(0, 0, this.camera.width, this.camera.height);
    this.context.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);

    for (String shaderId in instructions.keys) {
      Shader shader = this.resourceManager.getResource(ResourceManager.SHADER,
      resourceId:shaderId);

      shader.use(this.camera.projection);

      for (RenderInstruction instruction in this.instructions[shaderId]) {
        instruction.object.model.draw(instruction.matrix);
      }
    }

  }
}

